File: Recursion/8. Combinations/combinations.cpp
//https://leetcode.com/problems/combinations

class Solution {
public:
    void helper(int n, int k, int i, vector<int>& curr, vector<vector<int>>& res) {
        if(curr.size() == k) {
            res.push_back(curr);
            return;
        }
        if(i > n) return;

        //inc
        curr.push_back(i);
        helper(n, k, i+1, curr, res);
        curr.pop_back();

        //exc
        helper(n, k, i+1, curr, res);
    }

    vector<vector<int>> combine(int n, int k) {
        vector<int> curr;
        vector<vector<int>> res;

        helper(n, k, 1, curr, res);
        return res;
    }
};

File: Recursion/8. Combinations/combinations.py
# https//leetcode.com/problems/combinations

# Actaul problem: make combinations of n numbers.
# Sub problem: make combinations of n-1(top down) numbers or 1,2....n-1,n numbers(bottom-up).

# Bottom-up approach

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        currentCombination,result=[],[]
        self.makeCombinations(k,n,1,currentCombination,result)
        return result
    def makeCombinations(self,k:int,n:int,index:int,currentCombination: List[int],result: List[List[int]]):
        if len(currentCombination)==k:
            result.append(currentCombination.copy())
            return
        if index>n:
            return
        # include
        currentCombination.append(index)
        self.makeCombinations(k,n,index+1,currentCombination,result)
        # exclude
        currentCombination.pop()
        self.makeCombinations(k,n,index+1,currentCombination,result)

# Top down approach

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        currentCombination,result=[],[]
        self.makeCombinations(k,n,n,currentCombination,result)
        return result
    def makeCombinations(self,k:int,n:int,index:int,currentCombination: List[int],result: List[List[int]]):
        if len(currentCombination)==k:
            result.append(currentCombination.copy())
            return
        if index<1:
            return
        # include
        currentCombination.append(index)
        self.makeCombinations(k,n,index-1,currentCombination,result)
        # exclude
        currentCombination.pop()
        self.makeCombinations(k,n,index-1,currentCombination,result)